### DISCUSSION TITLE
GTD first impressions and notes/questions.

### DISCUSSION THREAD

**Discussion by molleweide** [nvim-neorg/neorg#307] [General]
Hi guys,

First of all neorg and GTD is pretty amazing when one is starting to get used to it! Much love and appreciate!

Below I post some notes that I have scribbled down and thought about after having used GTD for a couple of days.
I have forked and cloned all neorg repos and am looking forward to helping. I hope my notes add value and aren't
too stupid.

## example workspace

the examples workspace should be populated with more stuff so that it is easier
for new users to explore eg gtd views as a first time user.  Personally I am kind of starting
to understand how the syntax of gtd and neorg comes together but a more complex example workspace
would be great for understanding. Especially since I am confused as to whether I have followed
the syntax and therefore I don't know if `gtd views` is capturing my tasks etc. correctly.
I will also try to populate the repo further myself.
    
## setup neorg dirs and files

- is it required to have a neorg dir created already? todo make sure all dirs and files are setup correctly.

1. neorg dir DOES EXIST       > neorg workspace default does create a default workspace with an index file.
2. neorg dir DOES EXIST       > neorg capture does create a gtd subdir with an index file in current buffer.
3. neorg dir does NOT exist   > neorg capture does not create a neorg dir and gtd subdir with an index/inbox file.

- One has to manually create `inbox.lua` for gtd commands to work it seems.

- in what locations does `gtd views` command look? Currently, I don't
  understand how this works.

- maybe gtd should setup all of the base dirs and files necessary for the user.
  maybe even some dummy tasks/projects? I am not sure what's best. 

## bugs

- aborting/closing capture window generates error when running capture command anew. i need to restart neovim

## cwd

- switching to `gtd` index file if so only by using telescope find file when cwd=~/neorg
cwd is auto switched to gtd and this is a bit annoying since it restricts file search to
only `gtd` dir. One needs to switch back to `~/neorg` explicitly to be able to search
files outside of `gtd` dir but within neorg workspace. Is a workspace always auto setting
`:cwd`? I am not sure what I want but I noticed this so I thought I'd mention it.

## misc

I tried to open neorg workspace from within webbrowser with `firenvim` and it seems to work lol.

**Comment by danymat** [comment]
Hello @molleweide, GTD developer here !

> First of all neorg and GTD is pretty amazing when one is starting to get used to it! Much love and appreciate!

Thanks a lot for your appreciation, it's very important for me to know that users benefit from this as well as me, so thanks !

> Example workspace

You're right, the example gtd workspace is too small to help people understand how GTD works in Neorg.
We definitely should improve this. 

( Note for myself: we should reference the example workspaces in the new readme rewrite: #284 )

> is it required to have a neorg dir created already? todo make sure all dirs and files are setup correctly.

You're right, the logic here is a bit messy and not prone to a quick understanding of what's going on. If I remember correctly, GTD needs to have an existing workspace set up. If no workspace is set up, it'll fallback to the `default` workspace. And default workspace is the last workspace you entered in Neorg.
I think we'll need to add a check for the Neorg workspace right when launching Neorg. It's doable and I added it to my todo !

> One has to manually create inbox.lua for gtd commands to work it seems.

Right. And I don't know if we should or not create the inbox.norg file for the user if it does not exist. What do you think ?

> in what locations does gtd views command look? Currently, I don't
understand how this works.

`gtd views` will find every norg file in your gtd workspace, except the inbox.norg file.
In fact, I'm asking myself whether we should remove the `inbox.norg` file. Why?

Because we can improve the views a little bit by: Adding a new view called `Unprocessed`.

It'll group every task or project that is not entirely processed. For a task to be entirely processed, it has to be visible. If it's a task: by having one or more contexts, if it's a project: by having one or more tasks associated.
In this way, we can remove the inbox file because we could view our unprocessed tasks, and we will not lose any tasks from the views.
In fact, I'm sure it'll be a better feature, because the inbox file is quite difficult to grasp for new users, and still not very usable by advanced users (because moving tasks and projects is not yet implemented in GTD actions).

And I'm planning to create an option in GTD views, called `More informations` (or any other name really), that'll tell the user what files where parsed, and how many tasks and projects where found in GTD workspace. I guess this should help with your point.

> maybe even some dummy tasks/projects? I am not sure what's best.

Now that you told me this, it's a very good idea indeed. I know that Things3, and Todoist ask users upon launching the app first time, if they want to have a dummy project to know what to do with their app. I guess we can do this, and I will add it to my GTD todo ;).

> aborting/closing capture window generates error when running capture command anew. i need to restart neovim

Can you create a proper issue for this bug, so I can track it down ?

> switching to gtd index file if so only by using telescope find file when cwd=~/neorg
cwd is auto switched to gtd and this is a bit annoying since it restricts file search to
only gtd dir. One needs to switch back to ~/neorg explicitly to be able to search
files outside of gtd dir but within neorg workspace. Is a workspace always auto setting
:cwd? I am not sure what I want but I noticed this so I thought I'd mention it.

I recently implemented a way to not triggering workspace detection when visiting gtd views. You can do so by launching Neorg in silent mode: `NeorgStart silent=true` and launching your GTD views: `:Neorg gtd views`.
I will add this explicitly to the wiki docs

Once again, thanks a lot for your feedback, some true gems spotted in here and some todos to keep me alive for the time being 😄 !

**Reply by molleweide** [reply]
> ( Note for myself: we should reference the example workspaces in the new readme rewrite: #284 )
> 
> >

good idea.

> I think we'll need to add a check for the Neorg workspace right when launching Neorg. It's doable and I added it to my todo !
> 
> >

nice. maybe neorg should have a check for sub-workspaces and have an option to always set CWD to the upper most parent workspace?



> `gtd views` will find every norg file in your gtd workspace, except the inbox.norg file. In fact, I'm asking myself whether we should remove the `inbox.norg` file. Why?
> 
> Because we can improve the views a little bit by: Adding a new view called `Unprocessed`.
> 
> It'll group every task or project that is not entirely processed. For a task to be entirely processed, it has to be visible. If it's a task: by having one or more contexts, if it's a project: by having one or more tasks associated. In this way, we can remove the inbox file because we could view our unprocessed tasks, and we will not lose any tasks from the views. In fact, I'm sure it'll be a better feature, because the inbox file is quite difficult to grasp for new users, and still not very usable by advanced users (because moving tasks and projects is not yet implemented in GTD actions).
> 
> And I'm planning to create an option in GTD views, called `More informations` (or any other name really), that'll tell the user what files where parsed, and how many tasks and projects where found in GTD workspace. I guess this should help with your point.

Great to hear that you are working on this (which I also knew since I am following the project :) ). Also, it is still difficult for me to comment on this since I don't have enought knowledge of how gtd should actually work that I have a good intuition to understand I feel (refering back to a more detailed example workspace..) The only thing I can say is that is is good to hear about your work and progress.

> Now that you told me this, it's a very good idea indeed. I know that Things3, and Todoist ask users upon launching the app first time, if they want to have a dummy project to know what to do with their app. I guess we can do this, and I will add it to my GTD todo ;).
> 
> >

I think this would be a tremendous idea. To ask user if they want to load a couple of useful demo projects (home/work/school/side-projects) and also I think that maybe if the example workspace could be made into a opinionated `neorg-gtd-life-organizer-template` repo. What do you think of that? My thinking is that over time this
workspace could grow into quite a complex one that could work for most people out of the box. In any case I don't think it is too much of a problem deleting some of the premade gtd stuff as long as one gets a true feel for how good gtd could be.

> Can you create a proper issue for this bug, so I can track it down ?
> 
> >

I asked under `octo.nvim` about a related issue (https://github.com/pwntester/octo.nvim/issues/240) and Alvaro told me that one shouldn't really use `<c-c>` to abort, and in most cases use `<esc>` since neovim doesn't handle `<c-c>` the same (which is obvious now that I think about it), but usually when <esc> doesn't work for abortin `<c-c>` usually works.. 
So in the end a feature request for gtd would be to have esc for aborting task. I can still create an issue if you like.

**Reply by danymat** [reply]
> So in the end a feature request for gtd would be to have esc for aborting task. I can still create an issue if you like.

I think this is already the case in Neorg GTD

**Reply by danymat** [reply]
> hink this would be a tremendous idea. To ask user if they want to load a couple of useful demo projects (home/work/school/side-projects) and also I think that maybe if the example workspace could be made into a opinionated neorg-gtd-life-organizer-template repo. What do you think of that? My thinking is that over time this
workspace could grow into quite a complex one that could work for most people out of the box. In any case I don't think it is too much of a problem deleting some of the premade gtd stuff as long as one gets a true feel for how good gtd could be.

I'm not sure about creating a template though. I was more hoping for a GTD project with tasks telling the users what's possible to do with GTD and its views

This is a screenshot of the starter project in Things3

<img width="1070" alt="Capture d’écran 2022-01-17 à 15 29 29" src="https://user-images.githubusercontent.com/5306901/149788122-580f0811-7f93-4b47-96d3-98d952a6e4a2.png">

**Reply by molleweide** [reply]
Yeah I might have been using the word template wrong here because I agree with you.

**Reply by molleweide** [reply]
Btw your native lang is french?

**Reply by molleweide** [reply]
Btw your native lang is french?

**Reply by danymat** [reply]
It is indeed 😄

**Reply by danymat** [reply]
I just created a tutorial project for GTD, available here: https://github.com/nvim-neorg/example_workspaces#neorg-gtd-tutorial

Let me know what you think !

**Reply by molleweide** [reply]
I will dude!! I'm really looking forward to it. i already cloned it so I'll pull the changes later today.

**Reply by molleweide** [reply]
Oh btw, I pushed a snippets lib plugin for LuaSnip the other day https://github.com/molleweide/LuaSnip-snippets.nvim.
I have started adding neorg snippets to the neorg file. Does this make sense to you?

**Reply by molleweide** [reply]
i have to learn more about how to create lua snippets because it is quite hard, but the end goal would be to create nice dynamic snippets for everything neorg. you can do some crazy things with LuaSnip that is way beyond me atm.

**Reply by molleweide** [reply]
I like the `gtd/index.norg` file a lot. It instantly shed light on a couple of things that have been unclear to me. So thanks for this!

**Reply by molleweide** [reply]
I am also going to read through the docs again with this workspace side by side and see if I can contribute something myself also now.

**Reply by danymat** [reply]
Nice, I'm happy that it's been useful for you ! I wanted to provide an "out of the box experience" for most users

Feel free to hop on the discord and ping me if you want to contribute to gtd !

**Reply by molleweide** [reply]
sure! I should do that. I've also had on my todo list to check out a couple of TUI programs for discord. Because I don't like how discord interfers with Surfingkeys. I actually get quite annoyed when I use their website so this is a good push for me to get it running in the terminal.

**Reply by danymat** [reply]
> Oh btw, I pushed a snippets lib plugin for LuaSnip the other day https://github.com/molleweide/LuaSnip-snippets.nvim.
> I have started adding neorg snippets to the neorg file. Does this make sense to you?

Nice ! I guess you can take a look at this pr: https://github.com/nvim-neorg/neorg/pull/255
It's a try to make snippets work in Neorg. One bug though, it seems that when a trigger starts with `@`, it'll not find the snippet. Maybe you can take a look, that would be quite awesome to use snippets directly in neorg

**Reply by molleweide** [reply]
Cool i'll try to check it out later! 

What do you mean by 'directly in neorg'? I guess it'll become clear when I read the PR..

**Comment by danymat** [comment]
> And I'm planning to create an option in GTD views, called More informations (or any other name really), that'll tell the user what files where parsed, and how many tasks and projects where found in GTD workspace. I guess this should help with your point.

I implemented this feature in latest version. You can do `:Neorg gtd views` and press `i`

**Reply by molleweide** [reply]
Ah cool that is a pretty neat feature.

**Comment by danymat** [comment]
> I think we'll need to add a check for the Neorg workspace right when launching Neorg. It's doable and I added it to my todo !

I did provide a more interesting way of doing this, in commit 449d76eb623dadf84b79864703190c4319c46b1b. 

- if there is an error when loading GTD, the commands will still be created, but they will fallback to an error message.

**Comment by danymat** [comment]
For reference: https://github.com/nvim-neorg/neorg/issues/695

**Comment by NTBBloodbath** [comment]
Hey, this is completely offtopic but recently a scam bot (content +18, wth github allowing this kind of stuff?) mentioned a lot of people in this thread with a link. I have deleted the comment, but I don't know if the notification remains so now you know the reason, my apologies to everyone who has been mentioned here, let's hope it doesn't happen again!
